### PR TITLE
Prevent Encrypted Repository from Closing IndexInput

### LINKED ISSUE #71326: [CI] EncryptedAzureBlobStoreRepositoryIntegTests testSnapshotWithLargeSegmentFiles failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/jryw2jr4nn5ko/tests/:x-pack:plugin:repository-encrypted:internalClusterTest/org.elasticsearch.repositories.encrypted.EncryptedAzureBlobStoreRepositoryIntegTests/testSnapshotWithLargeSegmentFiles

**Reproduction line:**
`./gradlew ':x-pack:plugin:repository-encrypted:internalClusterTest' --tests "org.elasticsearch.repositories.encrypted.EncryptedAzureBlobStoreRepositoryIntegTests.testSnapshotWithLargeSegmentFiles" -Dtests.seed=134EA8254BF4838C -Dtests.security.manager=true -Dtests.locale=zh-Hant-TW -Dtests.timezone=Antarctica/Vostok -Druntime.java=16 -Dtests.fips.enabled=true`

**Applicable branches:**
7.12

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.repositories.encrypted.EncryptedAzureBlobStoreRepositoryIntegTests&tests.test=testSnapshotWithLargeSegmentFiles

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: a value greater than <0>
     but: <0> was equal to <0>

  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.repositories.blobstore.ESBlobStoreRepositoryIntegTestCase.assertSuccessfulSnapshot(ESBlobStoreRepositoryIntegTestCase.java:497)
  at org.elasticsearch.repositories.blobstore.ESBlobStoreRepositoryIntegTestCase.assertSuccessfulSnapshot(ESBlobStoreRepositoryIntegTestCase.java:493)
  at org.elasticsearch.repositories.blobstore.ESMockAPIBasedRepositoryIntegTestCase.testSnapshotWithLargeSegmentFiles(ESMockAPIBasedRepositoryIntegTestCase.java:167)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:567)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:831)

```

### PR DESCRIPTION
**Pull request by original-brownbear**
We shouldn't allow the repo to close the original input stream ever. This is handled
further up-stream and not the responsibility of this code.

closes #71326 (apparently some combination of resetting can trigger a
close on the original stream which this change ensure will not happen any longer).

### CODE REVIEW

File: x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java [hunk 1]
```diff
@@ -637,7 +637,7 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize, b
                 InputStream encryptedInputStream = ChainingInputStream.chain(
                     dekIdBytes.streamInput(),
                     new EncryptionPacketsInputStream(
-                        inputStream,
+                        Streams.noCloseStream(inputStream),
```
Comment (original-brownbear) [inline x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:640]: I suppose we could also fix this in `EncryptionPacketsInputStream` but this seemed less trappy and easier to reason about for now.

**Comment by original-brownbear** [comment]
ping @albertzaharovits :)

**Review by albertzaharovits — APPROVED** [review]
I don't think this is the problem but I don't have a better suggestion.

Alternatively, do you think a different uploaded blob size (after marks/resets) can trigger this behavior?

**Comment by albertzaharovits** [comment]
Apologies for not reviewing sooner!

**Comment by original-brownbear** [comment]
Thanks @albertzaharovits 

> Alternatively, do you think a different uploaded blob size (after marks/resets) can trigger this behavior?

I looked into this and it seemed to me that the logic should work with the sizes we read from the stream before resetting. But yes, from reading the code it looks to me like reading too far ahead would trigger problems. Either way, we should under no circumstances close the given input stream because the API doesn't expect that and it could cause unexpected trouble down the line => Let's see how it behaves with this merged, we can reopen if it's still an issue and it should make it clearer where the issue with mark-reset lies if it doesn't fully fix things IMO :crossed_fingers:

**Pull request merged**